File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/FrameHandle/TimerItemBase.lua
--[[
    @brief 定时器，这个是不断增长的
]]

MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.DelayHandle.IDelayHandleItem");

local M = GlobalNS.Class(GlobalNS.IDelayHandleItem);
M.clsName = "TimerItemBase";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mInternal = 1;            -- 定时器间隔
    self.mTotalTime = 1;           -- 总共定时器时间
    self.mCurRunTime = 0;          -- 当前定时器运行的时间
    self.mCurCallTime = 0;         -- 当前定时器已经调用的时间
    self.mIsInfineLoop = false;     -- 是否是无限循环
    self.mIntervalLeftTime = 0;    -- 定时器调用间隔剩余时间
    self.mTimerDisp = GlobalNS.new(GlobalNS.TimerFunctionObject);         -- 定时器分发
    self.mIsDisposed = false;        -- 是否已经被释放
    self.mIsContinuous = false;     -- 是否是连续的定时器
end

function M:setFuncObject(pThis, func)
    self.mTimerDisp:setPThisAndHandle(pThis, func);
end

function M:setTotalTime(value)
    self.mTotalTime = value;
end

function M:getRunTime()
    return self.mCurRunTime;
end

function M:getCallTime()
    return self.mCurCallTime;
end

function M:getLeftRunTime()
    return self.mTotalTime - self.mCurRunTime;
end

function M:getLeftCallTime()
    return self.mTotalTime - self.mCurCallTime;
end

-- 在调用回调函数之前处理
function M:onPreCallBack()
    
end

function M:OnTimer(delta)
    if self.mIsDisposed then
        return;
    end

    self.mCurRunTime = self.mCurRunTime + delta;
    if(self.mCurRunTime > self.mTotalTime) then
        self.mCurRunTime = self.mTotalTime;
    end
    self.mIntervalLeftTime = self.mIntervalLeftTime + delta;

    if self.mIsInfineLoop then
        self:checkAndDisp();
    else
        if self.mCurRunTime >= self.mTotalTime then
            self:disposeAndDisp();
        else
            self:checkAndDisp();
        end
    end
end

function M:disposeAndDisp()
    if(self.mIsContinuous) then
        self:continueDisposeAndDisp();
    else
        self:discontinueDisposeAndDisp();
    end
end

function M:continueDisposeAndDisp()
    self.mIsDisposed = true;
    
    while (self.mIntervalLeftTime >= self.mInternal and self.mCurCallTime < self.mTotalTime) do
        self.mCurCallTime = self.mCurCallTime + self.mInternal;
        self.mIntervalLeftTime = self.mIntervalLeftTime - self.mInternal;
        self:onPreCallBack();

        if (self.mTimerDisp:isValid()) then
            self.mTimerDisp:call(self);
        end
    end
end

function M:discontinueDisposeAndDisp()
    self.mIsDisposed = true;
    self.mCurCallTime = self.mTotalTime;
    self:onPreCallBack();
    
    if (self.mTimerDisp:isValid()) then
        self.mTimerDisp:call(self);
    end
end

function M:checkAndDisp()
    if(self.mIsContinuous) then
        self:continueCheckAndDisp();
    else
        self:discontinueCheckAndDisp();
    end
end

-- 连续的定时器
function M:continueCheckAndDisp()
    while (self.mIntervalLeftTime >= self.mInternal) do
        self.mCurCallTime = self.mCurCallTime + self.mInternal;
        self.mIntervalLeftTime = self.mIntervalLeftTime - self.mInternal;
        self:onPreCallBack();

        if (self.mTimerDisp:isValid()) then
            self.mTimerDisp:call(self);
        end
    end
end

-- 不连续的定时器
function M:discontinueCheckAndDisp()
    if (self.mIntervalLeftTime >= self.mInternal) then
        self.mCurCallTime = self.mCurCallTime + ((math.floor(self.mIntervalLeftTime / self.mInternal)) * self.mInternal);
        self.mIntervalLeftTime = self.mIntervalLeftTime % self.mInternal;   -- 只保留余数
        self:onPreCallBack();

        if (self.mTimerDisp:isValid()) then
            self.mTimerDisp:call(self);
        end
    end
end

function M:reset()
    self.mCurRunTime = 0;
    self.mCurCallTime = 0;
    self.mIntervalLeftTime = 0;
    self.mIsDisposed = false;
end

function M:setClientDispose()
    
end

function M:getClientDispose()
    return false;
end

function M:Start(priority)
    GCtx.mTimerMgr:addTimer(self);
end

function M:Stop()
    GCtx.mTimerMgr:removeObject(self);
end

return M;

File: Project/MyProject/Tool/CSToolKit/Project/ToolKit/ToolKit/Source/Libs/FrameHandle/TimerItemBase.cs
namespace SDK.Lib
{
    /**
     * @brief 定时器，这个是不断增长的
     */
    public class TimerItemBase : GObject, IDispatchObject, INoOrPriorityObject
    {
        public float mInternal;        // 定时器间隔
        public float mTotalTime;       // 总共定时器时间
        public float mCurRunTime;      // 当前定时器运行的时间
        public float mCurCallTime;     // 当前定时器已经调用的时间
        public bool mIsInfineLoop;      // 是否是无限循环
        public float mIntervalLeftTime; // 定时器调用间隔剩余时间
        public TimerFunctionObject mTimerDispatch;       // 定时器分发
        public bool mIsDisposed;             // 是否已经被释放
        public bool mIsContinuous;         // 是否是连续的定时器

        public TimerItemBase()
        {
            this.mInternal = 1;
            this.mTotalTime = 1;
            this.mCurRunTime = 0;
            this.mCurCallTime = 0;
            this.mIsInfineLoop = false;
            this.mIntervalLeftTime = 0;
            this.mTimerDispatch = new TimerFunctionObject();
            this.mIsDisposed = false;
            this.mIsContinuous = false;
        }

        public void setFuncObject(ICalleeObject pThis, MEventDispatchAction<TimerItemBase> handle)
        {
            this.mTimerDispatch.setFuncObject(pThis, handle, 0);
        }

        virtual public void setTotalTime(float value)
        {
            this.mTotalTime = value;
        }

        virtual public float getRunTime()
        {
            return this.mCurRunTime;
        }

        virtual public float getCallTime()
        {
            return this.mCurCallTime;
        }

        virtual public float getLeftRunTime()
        {
            return this.mTotalTime - this.mCurRunTime;
        }

        virtual public float getLeftCallTime()
        {
            return this.mTotalTime - this.mCurCallTime;
        }

        // 在调用回调函数之前处理
        protected virtual void onPreCallBack()
        {

        }

        public virtual void OnTimer(float delta)
        {
            if (this.mIsDisposed)
            {
                return;
            }

            this.mCurRunTime += delta;

            if (this.mCurRunTime > this.mTotalTime)
            {
                this.mCurRunTime = this.mTotalTime;
            }
            this.mIntervalLeftTime += delta;

            if (this.mIsInfineLoop)
            {
                this.checkAndDispatch();
            }
            else
            {
                if (this.mCurRunTime >= this.mTotalTime)
                {
                    this.disposeAndDispatch();
                }
                else
                {
                    this.checkAndDispatch();
                }
            }
        }

        public virtual void disposeAndDispatch()
        {
            if (this.mIsContinuous)
            {
                this.continueDisposeAndDispatch();
            }
            else
            {
                this.discontinueDisposeAndDispatch();
            }
        }

        protected void continueDisposeAndDispatch()
        {
            this.mIsDisposed = true;

            while (this.mIntervalLeftTime >= this.mInternal && this.mCurCallTime < this.mTotalTime)
            {
                this.mCurCallTime = this.mCurCallTime + this.mInternal;
                this.mIntervalLeftTime = this.mIntervalLeftTime - this.mInternal;
                this.onPreCallBack();

                if (this.mTimerDispatch.isValid())
                {
                    this.mTimerDispatch.call(this);
                }
            }
        }

        protected void discontinueDisposeAndDispatch()
        {
            this.mIsDisposed = true;
            this.mCurCallTime = this.mTotalTime;
            this.onPreCallBack();

            if (this.mTimerDispatch.isValid())
            {
                this.mTimerDispatch.call(this);
            }
        }

        public virtual void checkAndDispatch()
        {
            if(this.mIsContinuous)
            {
                this.continueCheckAndDispatch();
            }
            else
            {
                this.discontinueCheckAndDispatch();
            }
        }

        // 连续的定时器
        protected void continueCheckAndDispatch()
        {
            while (this.mIntervalLeftTime >= this.mInternal)
            {
                // 这个地方 m_curCallTime 肯定会小于 m_totalTime，因为在调用这个函数的外部已经进行了判断
                this.mCurCallTime = this.mCurCallTime + this.mInternal;
                this.mIntervalLeftTime = this.mIntervalLeftTime - this.mInternal;
                this.onPreCallBack();

                if (this.mTimerDispatch.isValid())
                {
                    this.mTimerDispatch.call(this);
                }
            }
        }

        // 不连续的定时器
        protected void discontinueCheckAndDispatch()
        {
            if (this.mIntervalLeftTime >= this.mInternal)
            {
                // 这个地方 m_curCallTime 肯定会小于 m_totalTime，因为在调用这个函数的外部已经进行了判断
                this.mCurCallTime = this.mCurCallTime + (((int)(this.mIntervalLeftTime / this.mInternal)) * this.mInternal);
                this.mIntervalLeftTime = this.mIntervalLeftTime % this.mInternal;   // 只保留余数
                this.onPreCallBack();

                if (this.mTimerDispatch.isValid())
                {
                    this.mTimerDispatch.call(this);
                }
            }
        }

        public virtual void reset()
        {
            this.mCurRunTime = 0;
            this.mCurCallTime = 0;
            this.mIntervalLeftTime = 0;
            this.mIsDisposed = false;
        }

        public void startTimer()
        {
            Ctx.msInstance.mTimerMgr.addTimer(this);
        }

        public void stopTimer()
        {
            Ctx.msInstance.mTimerMgr.removeTimer(this);
        }
    }
}